ekncontent/ekncontent/eknc-domain.c

```diff
@@ -0,0 +1,866 @@
+      EosShardShardFile *shard = g_object_new (EOS_SHARD_TYPE_SHARD_FILE,
+                                               "path", g_file_get_path (subscription_shard_file),
+                                               NULL);
+      self->shards = g_slist_append (self->shards, shard);
+    }
+
+  if (!eknc_utils_parallel_init (self->shards, 0, cancellable, error))
+    return FALSE;
+
+  if (!eknc_domain_setup_link_tables (self, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+initable_iface_init (GInitableIface *initable_iface)
+{
+  initable_iface->init = eknc_domain_initable_init;
+}
+
+static EosShardRecord *
+eknc_domain_load_record_from_hash_sync (EkncDomain *self,
+                                        const gchar *hash)
+{
+  for (GSList *l = self->shards; l; l = l->next)
+    {
+      EosShardRecord *record = eos_shard_shard_file_find_record_by_hex_name (l->data, hash);
+      if (record != NULL)
+        return record;
+    }
+  return NULL;
+}
+
+static GHashTable *
+eknc_get_domain_query_params (EkncDomain *self)
+{
+  GHashTable *params = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, g_free);
+  g_hash_table_insert (params, "manifest_path", g_file_get_path (self->manifest_file));
+  return params;
+}
+
+/**
+ * eknc_domain_get_subscription_id:
+ * @self: the domain
+ *
+ * Gets subscription id of the domain
+ *
+ * Returns: (transfer none): the subscription id
+ */
+const gchar *
+eknc_domain_get_subscription_id (EkncDomain *self)
+{
+  g_return_val_if_fail (EKNC_IS_DOMAIN (self), NULL);
+
+  return self->subscription_id;
+}
+
+/**
+ * eknc_domain_get_subscription_entries:
+ * @self: the domain
+ *
+ * Gets the json node containing the list of subscriptions from the
+ * subscriptions file.
+ *
+ * Returns: (transfer none): the json node
+ */
+JsonNode *
+eknc_domain_get_subscription_entries (EkncDomain *self)
+{
+  g_return_val_if_fail (EKNC_IS_DOMAIN (self), NULL);
+
+  return json_object_get_member (json_node_get_object (self->subscriptions_node), "subscriptions");
+}
+
+/**
+ * eknc_domain_get_shards:
+ * @self: the domain
+ *
+ * Gets the list of shard files in the domain.
+ *
+ * Returns: (element-type EosShardShardFile) (transfer none): the shards
+ */
+GSList *
+eknc_domain_get_shards (EkncDomain *self)
+{
+  g_return_val_if_fail (EKNC_IS_DOMAIN (self), NULL);
+
+  return self->shards;
+}
+
+/**
+ * eknc_domain_test_link:
+ * @self: the domain
+ * @link: the ekn id of link to check for
+ * @error: #GError for error reporting.
+ *
+ * Attempts to determine if the given link corresponds to content within
+ * this domain.
+ *
+ * Returns: (transfer full): Returns an EKN URI to that content if so, and NULL otherwise.
+ */
+const gchar *
+eknc_domain_test_link (EkncDomain *self,
+                       const gchar *link,
+                       GError **error)
+{
+  for (GSList *l = self->link_tables; l; l = l->next)
+    {
+      gchar *resolved = eos_shard_dictionary_lookup_key (l->data, link, error);
```

Comment:
Ditto here, fixing this properly would require marking a whole bunch of eos-shard internal function `char*` parameters const in the bloom filter, among others